src/Components/Web.JS/src/Services/NavigationManager.ts

```diff
@@ -7,39 +7,82 @@ import { EventDelegator } from '../Rendering/Events/EventDelegator';
 
 let hasEnabledNavigationInterception = false;
 let hasRegisteredNavigationEventListeners = false;
+let hasLocationChangingEventListeners = false;
+let currentHistoryIndex = 0;
+let navigationPromptOptions: NavigationPromptOptions | null = null;
 
 // Will be initialized once someone registers
 let notifyLocationChangedCallback: ((uri: string, state: string | undefined, intercepted: boolean) => Promise<void>) | null = null;
+let notifyLocationChangingCallback: ((uri: string, intercepted: boolean) => Promise<boolean>) | null = null;
+
+let popStateCallback: ((state: PopStateEvent) => Promise<void> | void) = onBrowserInitiatedPopState;
 
 // These are the functions we're making available for invocation from .NET
 export const internalFunctions = {
   listenForNavigationEvents,
   enableNavigationInterception,
+  setHasLocationChangingListeners,
+  enableNavigationPrompt,
+  disableNavigationPrompt,
   navigateTo,
   getBaseURI: (): string => document.baseURI,
   getLocationHref: (): string => location.href,
 };
 
-function listenForNavigationEvents(callback: (uri: string, state: string | undefined, intercepted: boolean) => Promise<void>): void {
-  notifyLocationChangedCallback = callback;
+function listenForNavigationEvents(locationChangedCallback: (uri: string, state: string | undefined, intercepted: boolean) => Promise<void>, locationChangingCallback: (uri: string, intercepted: boolean) => Promise<boolean>): void {
+  notifyLocationChangedCallback = locationChangedCallback;
+  notifyLocationChangingCallback = locationChangingCallback;
 
   if (hasRegisteredNavigationEventListeners) {
     return;
   }
 
   hasRegisteredNavigationEventListeners = true;
-  window.addEventListener('popstate', () => notifyLocationChanged(false));
+  window.addEventListener('popstate', onPopState);
+  currentHistoryIndex = history.state?.index ?? 0;
 }
 
 function enableNavigationInterception(): void {
   hasEnabledNavigationInterception = true;
 }
 
+function setHasLocationChangingListeners(hasListeners: boolean) {
+  hasLocationChangingEventListeners = hasListeners;
+}
+
+function onBeforeUnload(event: BeforeUnloadEvent) {
+  event.preventDefault();
+  return event.returnValue = navigationPromptOptions?.message;
```

Comment:
Is this intended as shorthand for assignment and then return of the same value? If so would it be possible to split this into two statements for clarity?